[samuelneo] iP

samuelneo opened the pull request:
## Carbon

> * 1. - [x] This [sentence](https://en.wikipedia.org/wiki/Sentence_(linguistics)) _exists_ 🌱 to fulfil `formatting` requirements

```java
// and so does this code block
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

samuelneo committed:
Rename project to Carbon

samuelneo committed:
Add indentation to program output

samuelneo committed:
Implement echo

samuelneo committed:
Implement add and list

samuelneo committed:
Add Task class and refactor packages

samuelneo committed:
Implement mark and unmark as done

samuelneo committed:
Replace 'if' with lambda-style switch statement

samuelneo committed:
Add Todo, Event, Deadline

samuelneo committed:
Implement todo, event, deadline commands

samuelneo committed:
Remove redundant entries from .gitignore

samuelneo committed:
Add text UI tests

samuelneo committed:
Standardise text file extensions to lowercase .txt

samuelneo committed:
Add Exceptions and error handling

samuelneo committed:
Add deletion of tasks

samuelneo committed:
Improve error detection and update tests

samuelneo committed:
Refactor constants

samuelneo committed:
Implement loading tasks.txt

samuelneo committed:
Implement writing to tasks.txt

samuelneo committed:
Remove unused imports

samuelneo committed:
Refactor variables

samuelneo committed:
Merge branch 'branch-Level-7'

samuelneo committed:
Add datetime support

samuelneo committed:
Refactor DateTime class to Temporal

samuelneo committed:
Carbon.java: Fix import structure

samuelneo committed:
Merge branch 'branch-Level-8'

samuelneo committed:
Extract into Parser, Storage, TaskList, Ui

samuelneo committed:
Fix formatting inconsistency

samuelneo committed:
Merge branch 'branch-A-MoreOOP'

samuelneo committed:
Merge branch 'add-gradle-support'

samuelneo committed:
Add checkstyle

samuelneo committed:
Update build.gradle

samuelneo committed:
Merge branch 'branch-A-Gradle'

samuelneo committed:
Add tests

samuelneo committed:
Merge branch 'branch-A-JUnit'

samuelneo committed:
Update build.gradle

samuelneo committed:
Merge branch 'branch-A-Jar'

samuelneo committed:
Add JavaDoc comments

samuelneo committed:
Reorder imports

samuelneo committed:
Implement find command

samuelneo committed:
Merge branch 'branch-A-JavaDoc'

samuelneo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/carbon/task/TaskList.java
#	src/main/java/carbon/utils/Parser.java
#	src/main/java/carbon/utils/Storage.java
#	src/main/java/carbon/utils/Ui.java

samuelneo committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/carbon/task/TaskList.java

samuelneo committed:
Add missing punctuation

samuelneo committed:
Fix checkstyle for lambda-style switch statements

samuelneo committed:
Merge branch 'fix-case-checkstyle'

samuelneo committed:
Add JavaFX

samuelneo committed:
Extract to FXML

samuelneo committed:
Tweak GUI

samuelneo committed:
Port over text UI functionality to GUI

samuelneo committed:
Add command-specific colors and implement bye

samuelneo committed:
Clean up

samuelneo committed:
Merge branch 'branch-Level-10'

samuelneo committed:
Reorder imports

samuelneo committed:
Tweak colors

samuelneo committed:
Change DialogBox font family to Trebuchet MS

wzhua02 commented on src/main/java/carbon/task/Temporal.java:
```diff
@@ -0,0 +1,216 @@
+        TEXT
+    }
+
+    private static final DateTimeFormatter[] DATE_FORMATTERS = {
```
I like that you have a class dedicated to parse various formats of dates and time.

wzhua02 commented on src/main/java/carbon/task/Todo.java:
```diff
@@ -0,0 +1,15 @@
+     */
+    public Todo(String description) {
+        super('T', description);
+    }
```
Maybe can have the methods like toString and getStorageText for Todo objects implemented here? 
My thoughts are that the ToDo class seems redundant to have.

samuelneo replied:
maybe, but I don't really see a need to include the implementation here since `Todo` can just use the implementation in `Task`

wzhua02 reviewed (commented):
The code is well written, feel like I've learnt a lot just from reading your code.

samuelneo committed:
Parser.java: Tweak regex

samuelneo committed:
Add error message color and tweak colors

loyaltypollution commented on src/main/java/carbon/gui/MainWindow.java:
```diff
@@ -0,0 +1,72 @@
+        String input = userInput.getText();
+        String response = carbon.getResponse(input);
+        String command = carbon.getCommand();
+        dialogContainer.getChildren().addAll(
+                DialogBox.getUserDialog(input, userImage),
+                DialogBox.getCarbonDialog(response, carbonImage, command)
```
Good code split that is clear and in line with given conventions!

loyaltypollution commented on src/main/java/carbon/task/Temporal.java:
```diff
@@ -0,0 +1,216 @@
+ *     <li>Text (none of the above)</li>
+ * </ul>
+ */
+public class Temporal implements Comparable<Temporal> {
+    enum TemporalType {
+        DATE,
+        TIME,
+        DATETIME,
+        TEXT
+    }
```
Could I ask why the implementation of Temporal instead of the more encompassing java.time.LocalDateTime?

samuelneo replied:
good question, I wanted an object that could hold either a date, time, datetime, or text, so that it's more lenient for the user (e.g. the user can just specify a date instead of a datetime if the time isn't important)

loyaltypollution commented on src/main/java/carbon/utils/Parser.java:
```diff
@@ -0,0 +1,60 @@
+    public static Deadline parseDeadline(String text) {
+        // Regex matches a String of the form "{A} /by {B}", where A and B each contain at least
+        // one non-whitespace character
+        Matcher matcher = Pattern.compile("^\\s*(\\S.*?)\\s+/by\\s+(\\S.*?)\\s*$").matcher(text);
```
Perhaps the regex string could be made private static final in Parser?
This gives us an opportunity to name the regex too.

samuelneo replied:
yeah possibly, though I only use this regex once so I don't really see that as necessary (though if I end up using it again or extending my use of regex I'll probably do that)

loyaltypollution commented on src/main/java/carbon/utils/Parser.java:
```diff
@@ -0,0 +1,60 @@
+     * @param text Details of the Deadline.
+     * @return The Deadline object.
+     */
+    public static Deadline parseDeadline(String text) {
+        // Regex matches a String of the form "{A} /by {B}", where A and B each contain at least
+        // one non-whitespace character
+        Matcher matcher = Pattern.compile("^\\s*(\\S.*?)\\s+/by\\s+(\\S.*?)\\s*$").matcher(text);
+        if (!matcher.find()) {
+            throw new InvalidArgumentException(
+                    "Deadline commands should be formatted as \"deadline [description] /by [due date/time]\"");
+        }
+        return new Deadline(matcher.group(1), matcher.group(2));
+    }
```
However, using regex groups to parse the deadline directly is great!

loyaltypollution commented on src/main/java/carbon/utils/Storage.java:
```diff
@@ -0,0 +1,103 @@
+            try {
+                Scanner scanner = new Scanner(dataFile);
+                while (scanner.hasNextLine()) {
+                    String firstLine = scanner.nextLine();
+                    if (firstLine.isEmpty()) {
+                        break;
+                    }
+                    char type = firstLine.charAt(0);
+                    boolean isDone = scanner.nextLine().charAt(0) == '1';
+                    String description = scanner.nextLine().trim();
+                    Task task = switch (type) {
+                        case 'T' -> new Todo(description);
+                        case 'D' -> {
+                            String dueBy = scanner.nextLine().trim();
+                            yield new Deadline(description, dueBy);
+                        }
+                        case 'E' -> {
+                            String from = scanner.nextLine().trim();
+                            String to = scanner.nextLine().trim();
+                            yield new Event(description, from, to);
+                        }
+                        default -> throw new InvalidFileFormatException();
+                    };
+                    if (isDone) {
+                        task.markAsDone();
+                    }
+                    taskList.add(task);
```
Good job with the parsing of Storage file!

loyaltypollution commented on src/main/java/carbon/utils/Ui.java:
```diff
@@ -0,0 +1,77 @@
+     */
+    public Ui() {
+        this.taskList = new TaskList();
+        startMessage = Storage.loadDataFile(taskList) + "Hello! What can I do for you?";
```
For me, it is not very clear what startMessage refers to. I get confused between start of JVM initialization, start of GUI construction or start of chatbot.

Is welcomeMessage clearer?

samuelneo replied:
good point, I'll rename it to `welcomeMessage`

loyaltypollution commented on src/main/java/carbon/utils/Storage.java:
```diff
@@ -0,0 +1,103 @@
+        if (!dataFile.getParentFile().exists() && !dataFile.getParentFile().mkdirs()) {
+            throw new RuntimeException("Unable to create data/user/ directory");
+        }
+        boolean fileExists;
```
As per [convention given](https://se-education.org/guides/conventions/java/intermediate.html#variables), we could use verb+noun. 
```suggestion
        boolean hasFile;
```
Is "hasFile" clearer?

samuelneo replied:
nice catch, thanks! will correct it

loyaltypollution reviewed (commented):
LGTM. Just a few Nits to pick. Great job and amazing progress! 👍

samuelneo committed:
Refactor variables

samuelneo committed:
Add assertions

The current codebase does not utilise assertions, and gradle has
assertions disabled by default.

Assertions are useful to ensure the control flow and overall logic of
the program is as intended.

Let's add assertions to verify program logic, and update the
gradle configuration to enable assertions.

samuelneo committed:
Refactor miscellaneous sections

Several code sections fall short of code quality guidelines.

These sections may be less intuitive or readable to others.

Let's update these sections to improve the overall code quality.

samuelneo committed:
TaskList.java: Make Stream clearer

The Stream in the listTasks method of TaskList.java is closed early
due to a processing step that relies on the number of filtered
tasks.

Combining that step into the streaming process allows for more
readable code.

Let's refactor this section to improve readability.

samuelneo committed:
Merge pull request #1 from samuelneo/branch-A-Assertions

Add assertions

samuelneo committed:
Merge branch 'master' into branch-A-CodeQuality

samuelneo committed:
Merge branch 'master' into branch-A-Streams

samuelneo committed:
Merge pull request #2 from samuelneo/branch-A-CodeQuality

Refactor miscellaneous sections

samuelneo committed:
Merge branch 'master' into branch-A-Streams

samuelneo committed:
Merge pull request #3 from samuelneo/branch-A-Streams

TaskList.java: Make Stream clearer

samuelneo committed:
Add gradle.yml for CI

samuelneo committed:
Fix time format

The functionality of DateTimeFormatter is OS-dependent.

This causes inconsistencies when parsing times.

Let's fix this by specifying a Locale for DateTimeFormatter.

samuelneo committed:
Fix comment wording

samuelneo committed:
Add sort command

samuelneo committed:
Remove unused method

samuelneo committed:
Merge branch 'branch-C-Sort'

samuelneo committed:
Tweak personality aspects

samuelneo committed:
Merge branch 'branch-A-Personality'

samuelneo committed:
Add help command and change default window size

samuelneo committed:
Extract TemporalType

samuelneo committed:
Reorder fields

samuelneo committed:
Add warnings for various Temporal issues

samuelneo committed:
Add application icon and tweak message font size

samuelneo committed:
Add delete marked tasks feature

samuelneo committed:
Add window title

samuelneo committed:
Fix regression

samuelneo committed:
Add Ui.png

samuelneo committed:
Add image, description, and summary table to user guide

samuelneo committed:
Add command descriptions and hyperlinks

samuelneo committed:
Add getting started section and todo example

samuelneo committed:
Add examples for remaining commands

samuelneo committed:
Merge branch 'branch-A-UserGuide'

samuelneo committed:
Fix minor wording issues

samuelneo committed:
Tweak formatting to fix HTML issues

samuelneo committed:
Update root README.md

samuelneo committed:
Improve code quality

samuelneo committed:
Fix typo in event command format

samuelneo committed:
Fix scrolling

samuelneo committed:
Change Markdown processor to GFM

samuelneo committed:
Revert commit cfe849a

samuelneo committed:
Revert commit 3734a4f partially due to GFM issue

samuelneo committed:
Remove instruction to double-click jar from user guide
